src/infra/src/table/migration/m20241114_000001_create_folder_table.rs

```diff
@@ -0,0 +1,216 @@
+            .await?;
+        Ok(())
+    }
+
+    async fn down(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .drop_index(Index::drop().name(FOLDER_ORG_NAME_TYPE_IDX).to_owned())
+            .await?;
+        manager
+            .drop_index(Index::drop().name(FOLDER_FOLDER_ID_IDX).to_owned())
+            .await?;
+        manager
+            .drop_index(Index::drop().name(FOLDER_ORG_IDX).to_owned())
+            .await?;
+        manager
+            .drop_table(Table::drop().table(Folder::Table).to_owned())
+            .await?;
+        Ok(())
+    }
+}
+
+/// Statement to create the folder table.
+fn create_folder_table_statement() -> TableCreateStatement {
+    Table::create()
+        .table(Folder::Table)
+        .if_not_exists()
+        .col(
+            ColumnDef::new(Folder::Id)
+                .integer()
+                .not_null()
+                .auto_increment()
+                .primary_key(),
+        )
+        // Surrogate key for folders that at the time of writing is...
+        // - stored in meta.key2
+        // - set by the server in calls to CreateFolder
+        // - referenced by the client in calls to ListDashboards
+        .col(ColumnDef::new(Folder::FolderId).string_len(256).not_null())
+        .col(ColumnDef::new(Folder::Org).string_len(100).not_null())
+        .col(ColumnDef::new(Folder::Name).string_len(256).not_null())
+        .col(ColumnDef::new(Folder::Description).text())
+        // Folder type where...
+        // - 0 is a dashboards folder
+        // - 1 is an alerts folder
+        // - 2 is a reports folder
+        .col(ColumnDef::new(Folder::Type).small_integer().not_null())
+        .col(
+            ColumnDef::new(Folder::CreatedAt)
+                .timestamp()
+                .default(SimpleExpr::Keyword(Keyword::CurrentTimestamp))
+                .not_null(),
+        )
+        .to_owned()
+}
+
+/// Statement to create index on org.
+fn create_folder_org_idx_stmnt() -> IndexCreateStatement {
+    sea_query::Index::create()
+        .if_not_exists()
+        .name(FOLDER_ORG_IDX)
+        .table(Folder::Table)
+        .col(Folder::Org)
+        .to_owned()
+}
+
+/// Statement to create unique index on folder_id surrogate key.
+fn create_folder_folder_id_idx_stmnt() -> IndexCreateStatement {
+    sea_query::Index::create()
+        .if_not_exists()
+        .name(FOLDER_FOLDER_ID_IDX)
+        .table(Folder::Table)
+        .col(Folder::FolderId)
+        .unique()
+        .to_owned()
+}
+
+/// Statement to create unique index on org, name, type combination.
+fn create_folder_org_name_type_idx_stmnt() -> IndexCreateStatement {
+    sea_query::Index::create()
+        .if_not_exists()
+        .name(FOLDER_ORG_NAME_TYPE_IDX)
+        .table(Folder::Table)
+        .col(Folder::Org)
+        .col(Folder::Name)
+        .col(Folder::Type)
+        .unique()
+        .to_owned()
+}
```

Comment:
⚠️ TODO:

I just realized that if we aren't enforcing unique folder names in the `meta` table today then this constraint could be a problem. If an existing DB instance contains folders with duplicate names then the `populate_folder_table` migration will fail when we try to copy folders from the `meta` table into the new `folder` table.

I'm going to confirm whether or not we currently allow duplicate folder names in the `meta` table today. If we DO allow them then I will remove this index before merging.

Reply:
I removed this index since I don't see anywhere where we enforced unique folder names in the `meta` table.